common/quotas/limiter_test.go

```diff
@@ -149,6 +122,241 @@ func BenchmarkMultiStageRateLimiter1000Domains(b *testing.B) {
+	mockTime.Advance(time.Nanosecond)
+	assert.Equal(t, limiter.Limit(), rate.Limit(1))
+	assert.Equal(t, limiter.Limit(), rate.Limit(1))
+	// Offset the time from the TTL and ensure it's still respected
+	mockTime.Advance(1500 * time.Millisecond)
+	assert.Equal(t, limiter.Limit(), rate.Limit(2))
+	mockTime.Advance(time.Second - 1)
+	assert.Equal(t, limiter.Limit(), rate.Limit(2))
+	mockTime.Advance(time.Nanosecond)
+	assert.Equal(t, limiter.Limit(), rate.Limit(3))
+}
+
+func TestDynamicRateLimiter_Allow(t *testing.T) {
+	cases := []struct {
+		name       string
+		ttl        time.Duration
+		elapsed    time.Duration
+		initialRps int
+		newRps     int
+		expected   bool
+	}{
+		{
+			name:       "success",
+			ttl:        time.Second,
+			elapsed:    time.Second - 1,
+			initialRps: 1,
+			newRps:     0,
+			expected:   true,
+		},
+		{
+			name:       "success with update",
+			ttl:        time.Second,
+			elapsed:    time.Second - 1,
+			initialRps: 1,
+			newRps:     2,
+			expected:   true,
+		},
+		{
+			name:       "success only after update",
+			ttl:        time.Second,
+			elapsed:    time.Second,
+			initialRps: 0,
+			newRps:     1,
+			expected:   true,
+		},
+		{
+			name:       "fail",
+			ttl:        time.Second,
+			elapsed:    time.Second - 1,
+			initialRps: 0,
+			newRps:     1,
+			expected:   false,
+		},
+		{
+			name:       "fail with update",
+			ttl:        time.Second,
+			elapsed:    time.Second,
+			initialRps: 1,
+			newRps:     0,
+			expected:   false,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockTime := clock.NewMockedTimeSource()
+			rps := tc.initialRps
+			rpsFunc := func() float64 {
+				res := rps
+				rps = tc.newRps
+				return float64(res)
+			}
+			limiter := NewDynamicRateLimiter(rpsFunc, WithTimeSource(mockTime), WithTTL(tc.ttl), WithMinBurst(1))
+
+			mockTime.Advance(tc.elapsed)
+			res := limiter.Allow()
+			assert.Equal(t, tc.expected, res)
+		})
+	}
+}
+
+func TestDynamicRateLimiter_Limit(t *testing.T) {
+	cases := []struct {
+		name       string
+		ttl        time.Duration
+		elapsed    time.Duration
+		initialRps int
+		newRps     int
+		expected   rate.Limit
+	}{
+		{
+			name:       "success",
+			ttl:        time.Second,
+			elapsed:    time.Second - 1,
+			initialRps: 1,
+			newRps:     0,
+			expected:   rate.Limit(1),
+		},
+		{
+			name:       "refreshed",
+			ttl:        time.Second,
+			elapsed:    time.Second,
+			initialRps: 1,
+			newRps:     0,
+			expected:   rate.Limit(0),
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockTime := clock.NewMockedTimeSource()
+			rps := tc.initialRps
+			rpsFunc := func() float64 {
+				res := rps
+				rps = tc.newRps
+				return float64(res)
+			}
+			limiter := NewDynamicRateLimiter(rpsFunc, WithTimeSource(mockTime), WithTTL(tc.ttl), WithMinBurst(1))
+
+			mockTime.Advance(tc.elapsed)
+			res := limiter.Limit()
+			assert.Equal(t, tc.expected, res)
+		})
+	}
```

Comment:
this is essentially "update changes the limit", yea?

so it's
```go
rps := 0
rpsFunc := func() float64 {
	return float64(rps)
}
limiter := NewDynamicRateLimiter(rpsFunc, WithTimeSource(mockTime), WithTTL(time.Second), WithMinBurst(1))
assert.Equal(t, limiter.Limit(), 0)
mockTime.Advance(time.Second)
rps = 1
assert.Equal(t, limiter.Limit(), 1)
```
but with loooots of boilerplate

Reply:
Done, I think that's a good simplification.